src/common/core.rs

```diff
@@ -50,10 +50,15 @@ impl Drop for EventLoop {
     }
 }
 
-pub fn spawn_event_loop(
+/// Spawns event loop in a separate thread and returns a handle to communicate with it.
+pub fn spawn_event_loop<T: 'static, F>(
     token_counter_start: usize,
     event_loop_id: Option<&str>,
-) -> Result<EventLoop> {
+    init_user_data: F,
+) -> Result<EventLoop<T>>
+where
+    F: 'static + FnOnce() -> Option<T> + Send,
```

Comment:
I should have put a comment there :) I will, but to answer your question:

1. it's a function not an instance of T so that the user data would be instantiated in the same thread were `Core` is owned. That allows us to store `Rc` data in `Core`. Otherwise the compilation would fail because of missing `Send` trait for the `T`.
2. it returns `Option` instead of `T` because user data initialization can fail, e.g. when user data is bootstrap cache, it's construction might fail, if given file name does not exist, etc. I could alternatively return an error, but since user data is so generic, there's no way to define an error, so instead I'm returning an `Option` and `init_user_data()` can log the error before returning `None`.

Reply:
1. aha :bulb: neat ! :smile: 
2. hmm for this i would say `Option` is too much of knowledge for this abstract module. See the [reasoning and suggestion here](https://github.com/maidsafe/crust/pull/1092#discussion_r242879007)